torch_xla/experimental/xla_sharding.py

```diff
@@ -30,46 +26,44 @@ def mark_sharding(t: Union[torch.Tensor,
         >> assert np.prod(mesh_shape) == xm.xrt_world_size()
         >> partition_spec = (0, None)
         >> assert len(input.shape) == len(partition_spec)
-
     Examples
     —------------------------------
     mesh_shape = (4, 2)
     input = torch.randn(8, 32).to(xm.xla_device())
     # 4-way data parallel
     input = xs.mark_sharding(input, mesh_shape, (0, None))
-
     linear = nn.Linear(32, 10).to(xm.xla_device())
     # 2-way model parallel
     linear.weight = xs.mark_sharding(linear.weight, device_mesh, (None, 1))
-
     output = linear(input)
     # full replication
     output = xs.mark_sharding(output, device_mesh, (None, None))
-    """
+  """
   num_devices = xm.xrt_world_size()
   assert np.prod(mesh_shape) == num_devices, \
     f"{mesh_shape} is not mappable over {num_devices} devices."
   assert all((d >= 0 and d < len(mesh_shape)) for d in partition_spec if d), \
     f"partition_spec ({partition_spec}) contains out of bound index into mesh_shape."
-  # We might allow {len(partition_spec)} <= {len(t.shape)},
-  # where the unspecified ranks are replicated.
+  # TODO(yeounoh) allow unspecified ranks (len(partition_spec) <= len(t.shape)),
+  # for replication. For now, all input rank sharding should be specified.
   assert len(t.shape) == len(partition_spec), \
     f"Partition spec length ({len(partition_spec)}) is not equal to the input rank ({len(t.shape)})."
 
   device_ids = np.array(range(num_devices))
-  tile_assignment = list(device_ids.reshape(mesh_shape))
+  tile_assignment = device_ids.reshape(mesh_shape).tolist()
 
-  manual = False
-  replicated = False
+  manual, replicated, partial = False, False, False
   if all(d is None for d in partition_spec):
-    # TODO(yeounoh) support partial replication
     replicated = True
+  elif any(d is None for d in partition_spec):
+    partial = True
+
+  # TODO(yeounoh) suport partially replicated sharding.
+  assert not partial, "Partial replication is currently not supported."
 
   if isinstance(t, XLAShardedTensor):
-    # Update sharding annotation
     torch_xla._XLAC._xla_mark_sharding(t.global_tensor, tile_assignment,
```

Comment:
per our discussion, we might want to error out here and ask user to manully clear sharding which will move data back to cpu.